### ISSUE TITLE
ssh transport: connecting side leaks ssh zombie processes if server is not available

### ISSUE THREAD

**Issue by MoritzFago** [zrepl/zrepl#322]
Found my freenas rather unresponsive with >12.000 processes, most of them zombies (<defunct>). They all were descendants of the zrepl deamon.

```yml

jobs:
  - name: namem
    type: pull
    connect:
      type: ssh+stdinserver
      host: hostname
      user: root
      port: 22
      identity_file: /identity
    interval: 10m
    root_fs: "something"

    pruning:
      keep_sender:
        - type: not_replicated
         # keep all snapshots that were not created by zrepl
#        - type: regex
#          negate: true
#          regex: "^zrepl_.*"
        - type: grid
          grid: 1x1h(keep=all)
          regex: "zrepl_.*"
      keep_receiver:
        - type: grid
          grid: 1x1h(keep=all) | 24x1h | 35x1d | 6x30d
          regex: "zrepl_.*"
        - type: regex
          negate: true
          regex: "^zrepl_.*"
  - name: othername
    type: pull
    connect:
      type: ssh+stdinserver
      host: otherhost
      user: root
      port: 22
      identity_file: file
    interval: 10m
    root_fs: "otherfs"
    pruning:
      keep_sender:
        - type: not_replicated
         # keep all snapshots that were not created by zrepl
        - type: regex
          negate: true
          regex: "^zrepl_.*"
      keep_receiver:
        - type: grid
          grid: 1x1h(keep=all) | 24x1h | 35x1d | 6x30d
          regex: "zrepl_.*"
        - type: regex
          negate: true
          regex: "^zrepl_.*"
```

freebsd 11.3
client: zrepl version=v0.2.1 GOOS=freebsd GOARCH=amd64 Compiler=gc

**Comment by problame** [comment]
Were the processes `zfs` or `zrepl stdinserver` ? And were they on the sending or receiving side?

**Comment by MoritzFago** [comment]
The processes were on the pulling side. They were neither of the suggested, but ssh processes.

**Comment by problame** [comment]
Fix is in  b86bbda for master.
I won't issue a 0.2.2 release since 0.3 is practically around the corner.
However, I pushed a back-port of the patch in https://github.com/zrepl/zrepl/tree/problame/322-fix-ssh-client-leaks-backport-0.2.1 commit cd47103669686da22e544fe0af744632f24db324 

The CI should produce binary artifacts ready for download by clicking on the GitHub commit status:
![image](https://user-images.githubusercontent.com/956573/83411736-1adc3c00-a419-11ea-8092-83e8a187e68b.png)

You can overwrite the `pkg`-installed binaries with those binaries from the CI, chmod +x them, should be good to go. Please report both negative and positive outcomes ASAP so we can include the patch in 0.3.

**Comment by MoritzFago** [comment]
It‘s better now, but if I kill the zrepl daemon process, some of the ssh processes get reassigned init as parent process

**Comment by problame** [comment]
How are you killing the `zrepl daemon` process exactly?

**Comment by MoritzFago** [comment]
`pkill zrepl` acording to the man page, that sends a `TERM` signal.

**Comment by problame** [comment]
Are the orphaned ssh processes in zombie-state or are there still zrepl processes?
Init should reap them if they are zombies, and if they are still alive, they should terminated once the other side runs into the heartbeat timeout (30s / 1min IIRC).

**Comment by problame** [comment]
@MoritzFago any updates wrt above question?
ISTM that the initial fix posted above solved the most pressing issue with the ssh transport.

**Comment by problame** [comment]
Closing this due to inactivity.

**Comment by MoritzFago** [comment]
zrepl seems to leak memory. The memory seems to be unused, as it will slowly fill the swap-space

**Comment by problame** [comment]
So no leaked ssh processes anymore in the `ps aux` output, but the zrepl daemon consumes more memory?

(Please don't restart the daemon in that case, wen can capture a heap profile using proof.)

**Comment by problame** [comment]
Any updates on the, @MoritzFago ?

**Comment by MoritzFago** [comment]
I captured a heap profile today, and looked at the data. I limited the resource usage `sudo rctl -a user:zrepl:swapuse:deny=120M` I am unsure about the contents of a heap profile, so i don't want to upload it. Most of the space seems to be used by `NewCircularLog`. This is the output of: `http://localhost:8081/ui/top?si=inuse_space`, I captured the profile with ` curl -sK -v http://localhost:8888/debug/pprof/heap` 
<img width="818" alt="Bildschirmfoto 2020-09-27 um 03 50 41" src="https://user-images.githubusercontent.com/4459068/94353747-ee59c900-0074-11eb-9720-e859f036de14.png">
